### ISSUE TITLE
Add Object type support in Metadata processor

### ISSUE THREAD

**Issue by suchostone** [redux-autoform/redux-autoform#67]
Hi guys!  Can you pls evaluate supporting objects inside the form?

e.g.:

``` javascript
{
    a: String,
    foo: { 
        x: String,
        y: String
    } 
}
```

Then in autoform I should be able in my fields to do something like below:

``` javascript
{
    "displayName": "This x from foo",
    "name": "foo.x"
}
```

Thank you!

**Comment by suchostone** [comment]
possible approach, similar as type: "array" already supported:

```
{
      "name": “foo",
      "type": “entity",
       "entityType": “foo",
       "layoutName": "edit"
}
```

**Comment by andrerpena** [comment]
@suchostone , I totally agree. Actually, these are the data-types I wanted to support and the current state:
- [ ] Entity
- [X] Array of Entity
- [X] Primitive
- [ ] Array of Primitive

You're asking the first one. This shouldn't be difficult to do and I'm the best one to do it because I did the metadata processor. I'll take a look into it.

Thanks for the proposal.

**Comment by JonatanSalas** [comment]
@andrerpena could you take a look at this one, please? :smile:

**Comment by devang-kredx** [comment]
Is there a work around available to do it today? The objects in the form I am creating are quite deep and would really like to be able to use autoform in the state it is in today even though it might need more work right now :)

**Comment by andrerpena** [comment]
I'll try to solve this one tonight. Don't worry =)

**Comment by devang-kredx** [comment]
Just in case someone is stuck on this issue and wants to start using autoforms **today**, a hack I am using to support nested objects is to flatten it, use that to create the auto forms and populate initial values, and unflatten it when using the values to call the api. Code below to flatten/unflatten, modified slightly from [here](http://stackoverflow.com/a/19101235/712476) to make it compatible with autoform

```
const separator = '__'

// Converts all flat structures into nested ones
function unflatten(values) {
  "use strict";
  if (Object(data) !== data || Array.isArray(data))
      return data;
  var result = {}, cur, prop, idx, last, temp;
  for(var p in data) {
      cur = result, prop = "", last = 0;
      do {
          idx = p.indexOf(separator, last);
          temp = p.substring(last, idx !== -1 ? idx : undefined);
          cur = cur[prop] || (cur[prop] = (!isNaN(parseInt(temp)) ? [] : {}));
          prop = temp;
          last = idx + 1;
      } while(idx >= 0);
      cur[prop] = data[p];
  }
  return result[""];
}

// Converts all nested structures into flat files
function flatten(data) {
  var result = {};
  function recurse (cur, prop) {
      if (Object(cur) !== cur) {
          result[prop] = cur;
      } else if (Array.isArray(cur)) {
           for(var i=0, l=cur.length; i<l; i++)
               recurse(cur[i], prop ? prop+separator+i : ""+i);
          if (l == 0)
              result[prop] = [];
      } else {
          var isEmpty = true;
          for (var p in cur) {
              isEmpty = false;
              recurse(cur[p], prop ? prop+separator+p : p);
          }
          if (isEmpty)
              result[prop] = {};
      }
  }
  recurse(data, "");
  return result;
}

export {flatten, unflatten}
```

**Comment by JonatanSalas** [comment]
This is a really good idea! Thanks for your work around!

**Comment by andrerpena** [comment]
Thanks for this workaround. I'll implement this ASAP. It's just not being
easy for me to find time. You can expect a solution until the end of this
week.

Em 15 de set de 2016 11:21, "Jonatan E. Salas" notifications@github.com
escreveu:

**Comment by andrerpena** [comment]
I'm looking at the current codebase and it seems that the `entity` type is supported already, from the metadata processing stand point. I just think we don't have any UI component. I'll figure this out today.

**Comment by andrerpena** [comment]
I think I got it covered: https://redux-autoform.github.io/redux-autoform-bootstrap-ui/demo.html?preset=EntityContainer

The fix was on the BS-UI version 1.7.1-experimental

If something doesn't work as expected, feel free to reopen it.

**Comment by devang-kredx** [comment]
Entity support seems to work, but am not able to layout the entity fields in a horizontal manner. For example, if in the [demo](https://redux-autoform.github.io/redux-autoform-bootstrap-ui/demo.html?preset=EntityContainer), I want to have the `type` and `address` fields to be in the same row (size 6), I am unable to do that.

**Comment by andrerpena** [comment]
Thanks for the feedback. I'm reopening and will take a look ASAP.
